scripts/service.py

```diff
@@ -0,0 +1,286 @@
+# client side invoke, server side just do one invoke of the executable (assume the 
+# script do not need to be executed multiple times in case of multiple trigger). For
+# client side, it is just simple logic to open the same service pipe as server, and write
+# something to wake up the server to do the work.
+#
+# There are multiple input parameters,
+#    -s/--service:  this is the service name for both server and client, supposing a short name
+#    -m/--mode: this is the mode specify 'client' or 'server'
+#    -e/--executable: this is the executable script to be invoked upon server wake up, 
+#                     only needed for server logic. Arguments have to be included along with
+#                     script name as one string.
+#
+# Example:
+#    server:
+#       service.py -s usbscan -m server -e /opt/xensource/libexec/usb_change
+#
+#    client:
+#        service.py -s usbscan -m client
+#
+
+import argparse
+import os
+import select
+import subprocess
+import logging
+import xcp.logger as logger
+import errno
+import shutil
+import stat
+import sys
+
+BUFFER_SIZE = 2048
+
+class ServiceSkeleton(object):
+    """ base class for both client and server, maily as utilities """
+
+    PIPE_BASE_PATH = '/var/lib/misc/'
+    
+    def __init__(self, name):
+        super(ServiceSkeleton, self).__init__()
+        self.name = name
+        self.pipePath = self.PIPE_BASE_PATH + name
+        self.logPrefix = name + ":"
+        self._checkPath()
+        self.fifo = None
+
+    def _checkPath(self):
+        """ check if the paths and files exist, otherwise create them """
+        
+        if not os.path.exists(self.PIPE_BASE_PATH):
+            try:
+                os.mkdir(self.PIPE_BASE_PATH, 0755)
+            except OSError as err:
+                self.logErr("create base path failed!")
+                raise
+        
+        create_fifo = False
+        if os.path.exists(self.pipePath):
+            if not stat.S_ISFIFO(os.stat(self.pipePath).st_mode):
+                # as this is a specific file in our specific dir,
+                # assume it is used exclusively, so the current none pipe
+                # stuff is considered garbage and try to delete that here.
+                if os.path.isfile(self.pipePath):
+                    os.remove(self.pipePath)  # remove the file
+                    self.logInfo("removed file:" + self.pipePath)
+                elif os.path.isdir(self.pipePath):
+                    shutil.rmtree(self.pipePath)  # remove dir and all contains
+                    self.logInfo("removed dir:" + self.pipePath)
+                else:
+                    self.logErr("Unexpected file node at:" + self.pipePath)
+                    raise OSError(errno.EEXIST, "Unexpected file node exists", self.pipePath)
+
+                create_fifo = True
+        else:
+                create_fifo = True
+
+        if create_fifo:
+            try:
+                os.mkfifo(self.pipePath)
```

Comment:
Actually no. The worst case is udev killing those tasks after waiting some time and other event handlers which udev would have run not being run until that time has expired, or missed entirely as no workers were available. I have observed udev having this behaviour. systemd-udevd has a maximum number of children it will run, whose default is going to end up at around 8 per GB of RAM in Dom0 (src/udev/udevd.c in systemd). We can't rely on not hitting it if we block our workers too long - we are not the only thing udev is handling. 

From pipe(7): "If a process attempts to write to a full pipe (see below), then write(2) blocks until sufficient data has been read from the pipe to allow the write to complete."

So we have two cases where the client can block:

1) On open() when the FIFO exists but the server is not running
2) On write() when the FIFO is full, which can happen if a large number of events occurs while the server is off running the worker script.

1) can be avoided fairly easily, either by having that open() return ENOENT unless the server will be listening very soon (I mean milliseconds), or by doing a non-blocking open()
2) can be avoided by setting the opened fd non-blocking and ignoring EAGAIN and EPIPE from the write.

Reply:
@TimSmithCtx , OK, in that case, we can make the client part none blocking when at open.
For case 2, it is indeed that the server will not block waiting the child process to finish, so I guess we just do not need to take care of it.